utils/extract.go

```diff
@@ -10,65 +10,69 @@ import (
 	"io"
 	"os"
 	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
 )
 
+func sanitize(path string) (string, error) {
+	path = filepath.Clean(path)
+
+	if strings.HasPrefix(path, "..") {
```

Comment:
Not that it would ever actually happen, but theoretically you could have a valid file named `..foo` which would fail to extract because of this check. Maybe include the path separator just in case?

We could also get rid of the `HasPrefix` check entirely by prepending the path separator to `path` before passing it to `Clean`. That way the path would always be rooted and `Clean` would handle all the sanitization.

Reply:
Thanks for this -- it ended up being slightly more complicated, since I didn't want to just collapse `../../file.ext` and `file.ext` in a tar file -- so I've opted to wholesale reject all relative paths, but building upon the `filepath.Clean` methodology you suggested. Let me know what you think!

Reply:
Good thinking, but this won't work on Windows because `Clean` replaces slashes with `Separator` and on Windows `Separator` is backslash 😕

Reply:
Ah, hmm. Any objections to just rejecting a header containing the string `../` anywhere then, instead of just looking at the start?

Reply:
It's a bit difficult to do handle this elegantly without missing any corner cases. You'd have to account for paths like:

```go
"foo/../bar"   // Basic case; invalid path
"foo/..\bar"   // Windows separator; invalid path
"foo/.../bar"  // Three dots; valid path
"foo../bar"    // Two dots part of dir name; valid path
".."           // Only dots; invalid path
```
And possibly more. I'd really prefer leveraging `Clean` so we don't have to go through all these ourselves.

There's [`filepath.ToSlash`](https://golang.org/pkg/path/filepath/#ToSlash) which should help with the Windows separator issue. Maybe just add that to the current approach?

Reply:
That /should/ work, but even my current approach is starting to seem too complex. GNU Tar has the following implementation of [contains_dot_dot](https://git.savannah.gnu.org/cgit/tar.git/tree/src/names.c#n1954):
```c
/* Return nonzero if NAME contains ".." as a file name component.  */
bool
contains_dot_dot (char const *name)
{
  char const *p = name + FILE_SYSTEM_PREFIX_LEN (name);

  for (;; p++)
    {
      if (p[0] == '.' && p[1] == '.' && (ISSLASH (p[2]) || !p[2]))
	return 1;

      while (! ISSLASH (*p))
	{
	  if (! *p++)
	    return 0;
	}
    }
}
```

It [uses](https://git.savannah.gnu.org/cgit/tar.git/tree/src/extract.c#n1743) that during extract to reject as we do. A Go implementation of same would look like:
```
	if strings.Contains(path, ".."+string(filepath.Separator)) || strings.HasSuffix(path, "..") {
		return "", errors.Errorf("unexpected relative path %s", path)
	}
```

I believe it would reject `"foo../bar"` as invalid even though it might be a valid directory name. Thoughts on adopting this approach?

Reply:
Tested locally, GNU tar extracts `foo../bar` just fine.

What the C code does is looks for `../` or `..<nil>` and returns 1 if it sees that. But if it doesn't, it traverses to the next `/` and starts again, so it's always matching against the whole component and not just a suffix.

Also, comparing against `filepath.Separator` is the wrong thing to do here, since on Windows that's `\`, but tar files would mostly contain forward slashes. There's [`os.IsPathSeparator`](https://golang.org/pkg/os/#IsPathSeparator) that accounts for both, but that can't be easily used with `strings.Contains`.

Reply:
Ah, you’re right. I didn’t carefully examine the starting conditions for subsequent iterations of the for loop. I’ll take a stab at `filepath.ToSlash`.

Reply:
One other note: this code is used exclusively for unpacking plugins. Regardless of tackling the Windows issue, we do have the option of restricting our support for what kind of tar headers we’re wiling to unpack, and relocating this code to avoid implying otherwise.

Reply:
Relocating the code sounds like a good idea regardless of the final implementation we end up with. It's not a generic solution in any case since there's already lots of things that are explicitly not supported: links, file modes, owners, timestamps...

Reply:
Ok, I've pushed up the simplifications -- thanks for the feedback throughout all the thrashing :)